README.md
# dodeca-dragons

changelog.txt
v0.5.1 (25/10/22)
-Fixed issue where certain early game players would have platinum upgrade 7 at 20/10 instead of 10/10
-Swapped the rewards for cyan sigil achievements 1 and 2, meaning automatic spending time comes earlier
-"Gain more platinum and uranium upgrades" cost has been reduced from e450 to e420
-Achievement info no longer scrolls with the achievement icons, keeping it readable if you scroll down

v0.5.0 (23/10/22)
-Added achievements, which can give you QoL bonuses
-The QoL bonuses now work differently! Most now come from the new achievements instead of upgrades. They have also been generally moved earlier in order to make the game a bit less grindy
-Reworked several magic upgrades to rebalance for achievements
-Removed QoL related sigil upgrades and uranium upgrades as they have now been replaced with achievements
-You can now name your dragon! It doesn't have any practical effect, but it's cool
-Some sigil requirements and upgrade prices have been changed
-"Next magic at" text now persists for longer
-Other minor bug fixes and changes

v0.4.4 (10/10/22)
-Added pink sigils!
-All traces of "purple" sigils have been wiped from existence. We are not to speak of them any further.
-Fire max all now divides gold up equally between the upgrades instead of maxing each upgrade in a row
-Fixed performance issue in which render would occur too often when clicking and dragging
-Reduced "Unlock more magic upgrades" requirement from 275k magifolds to 150k
-Number formatting now has 2 digits of precision for the mantissa instead of 3
-Fixed some issues with loading earlier saves
-Dragging controls are now implemented for mobile
-Other minor bug fixes and changes

v0.4.3 (05/10/22)
-Added violet sigils!
-You can now pet the dragon (unlockable with violet sigils)
-Nerfed some of the indigo sigil upgrades
-The game now works on mobile!
-Added movement using the arrow keys
-Added option to disable the background pattern
-Some icons have been changed
-Fixed endgame inflation issue (magic effect is now softcapped again past e100,000)
-Fixed issue with clicking and dragging getting stuck if you drag your cursor out of the window
-Fixed issue with right clicks causing undesired panning
-Fixed fire upgrades resetting to the wrong price on magic reset
-Fixed more weird rounding issues
-Platinum upgrade 7 is now actually kept on magic reset like it says
-Fire upgrade 3 has been buffed slightly
-Began splitting script.js into multiple files for ease of reading going forward
-Other minor bug fixes and changes

v0.4.2 (30/09/22)
-Added indigo sigils!
-There's finally an upgrade to gain score automatically! YIPPEE
-Blue sigil effect is slightly stronger
-Fixed weird rounding issues for large powers of 10
-Changed number formatting for numbers over 1e10,000
-Changed magic and sigil resets so that you will always be asked to confirm if you will gain nothing from resetting
-Hopefully finally fixed the issue with dragging getting stuck on if your mouse is over a disabled button
-Other minor bug fixes and changes

v0.4.1 (27/09/22)
-Added blue sigils!
-Added a time played stat (in the settings and info tab)
-Fixed exploit that allowed players to gain lots of free gold at the start of challenges
-Fire effect text in dragon tab now displays correctly if you have dark magic upgrade 6
-Other minor bug fixes and changes

v0.4.0 (25/09/22)
-Added a new prestige layer: sigils!
-Dragon time spent effect now says "capped" when you hit the cap (^2.5)
-Magic effect is now softcapped past 1e1,000
-Rearranged a few tabs (settings tab is now in the top left corner with the resources tab)
-Screen now pops up for mobile users telling them they are unable to play on mobile (I haven't yet been able to get mobile to work)
-Even more buttons are no longer fuzzy
-Other minor bug fixes and changes

v0.3.2 (18/09/22)
-Added dark magic upgrades!
-Fixed fire buy max buttons not appearing when upgrade is bought
-Made the magic upgrade code a LOT shorter and faster

v0.3.1 (17/09/22)
-Added some new platinum and uranium upgrades!
-Added buy max buttons for fire upgrades
-Replaced the home icon with a better image
-More buttons are no longer fuzzy (yay!)
-Other minor bug fixes and changes

v0.3.0 (16/09/22)
-Added a new alchemy element: uranium!
-Added a new dragon stage with some fancy new features (Surprise! Your dragon has feelings)
-Made all the buttons left-aligned to get rid of that annoying unfixable subpixel rendering antialiasing crap
-Magic tab now shows the amount of gold needed for the next magic (as long as you don't have any magic-increasing upgrades)
-Fixed bug with max all for fire upgrades
-Fixed bug with fire upgrades 4 and 5
-Other minor bug fixes and changes

v0.2.3 (11/09/22)
-Added 2 new magic upgrades
-Added confirmation popup toggle for magic resets
-Importing earlier saves now makes later stuff disappear
-Fixed fire resource text not updating until fire upgrades are unlocked
-Some small balances to make early game easier

v0.2.2 (10/09/22)
-Various bug fixes

v0.2.1 (08/09/22)
-Fixed a bug where the game is completely broken and unplayable lol

v0.2.0 (08/09/22)
-Added magic challenges
-Added manifolds, which boost gold gain
-Added some new magic upgrades

v0.1.2 (06/09/22)
-Fixed bug where importing a save wouldn't hide max all buttons
-Added discord and changelog links

v0.1.1 (04/09/22)
-Added 4 new magic upgrades and a new platinum upgrade
-Main tab now display gold/click
-Fixed fire upgrade 2 not affecting production
-Fixed fire upgrade 3 being slightly incorrect
-Platinum upgrade 6 has been made slightly less powerful
-Added some new info text

v0.1.0 (04/09/22)
-Added magic (lose all resources and upgrades for a big bonus!)
-Added magic upgrades
-Added new dragon stage (dark dragon)
-Updated fire texture to better fit with the rest of the icons

v0.0.7 (30/08/22)
-Finished the rest of the platinum upgrades
-Added new dragon stage (elder dragon)
-Minor bug fixes

v0.0.6 (29/08/22)

constants.js
//const arrays aren't actually constant but shhh
//I am constantly suffering
const romanNumerals = ["I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII", "XIII", "XIV", "XV"]

const fireUpgradeInitialCosts = [0, 50, 100, 100, 500, 500, 2e7]
const fireUpgradeBase = [0, 1.8, 2, 1.4, 1.5, 2.5, 5]
const magicUpgradeCosts = [2, 3, 8, 12, 30, 100, 300, 1500, 4000, 20000, 100000, 400000, 1e8, 3e8, 1.5e9, 3e9, 1e11, 4e12, 1e13, 1e16]
const darkMagicUpgradeCosts = ["e110", "e125", "e165", "e240", "e285", "e320", "e360", "e450", "e125000", "e150000"]
const sigilColours = ["cyan", "blue", "indigo", "violet", "pink"]
const knowledgeMultipliers = [450, 112.5, 18.75, 10, 1]
const tomeUpgradeCosts = [1, 1, 2, 2, 4]
const achievementNames = [
  //Gold
  ["Starting strong", "Growing stronger", "The heap", "Motherlode", "BEZOS BEZOS BEZOS", "Unending mound of cash", "Multiversal", "Endless realm of coins", "Loaded", "Mega-loaded (???)", "Sbiisics", "On my dime!"],
  //Fire
  ["Spicy", "Burning hands", "The great fire", "Towering inferno", "Hell pit", "Carolina reapers", "Tartarus", "Pandemonium", "Infernal overlord", "Unfortunate events"],
  //Platinum
  ["Alchemagoria", "Jeweller", "Putting in the work", "Magnum opus", "Simple chemistry", "Mirror polished", "Silver skyscrapers", "Advanced chemistry", "God of transmutation", "Platisynthesis"],
  //Magic
  ["Funny wizard hat", "Sorcery", "Magiccery!", "THE BIGGER HAT", "Thaumaturgy", "Spells galore", "Reshape the Earth", "Unlimited potential", "Star shatterer", "Endless grimoires", "Void tapping", "Threshold memories"],
  //Magifolds
  ["4-dimensional", "10-dimensional", "Superstring theory", "Infinitely folding", "Incomprehensible", "Megafolds", "GIGAFOLDS", "Spacetime origami", "Nonagon Infinity"],
  //Uranium
  ["Ionizing", "Radiating", "Neutralizing", "Terrifying", "Devastating", "Decaying", "Annihilating"],
  //Cyan sigils
  ["Chromatic", "Chaotic", "Colorific", "Catastrophic", "Catatonic", "Chromatic II", "Chaotic II"],
  //Blue sigils
  ["Bichromatic", "Basic", "Biologic", "Barbaric", "Balsamic", "Bichromatic II", "Basic II"],
  //Indigo sigils
  ["Interdynamic", "Idealistic", "Isotropic", "Isotonic", "Idyllic", "Interdynamic II", "Idealistic II"],
  //Violet sigils
  ["Volcanic", "Volumetric", "Vitriolic", "Voltaic", "Vinic", "Volcanic II", "Volumetric II"],
  //Pink sigils
  ["Panchromatic", "Psychotic", "Pathogenic", "Particularistic", "Platonic", "Panchromatic II", "Psychotic II"],
  //Knowledge
  ["Big brain moment", "Deep thought", "Immaterialistic", "Planning ahead", "I conject"],
  //Tomes
  ["History locked away", "The mystical bookshelf"]
]
const achievementRequirements = [
  //Gold
  ["100", "10000", "1e6", "1e12", "1e20", "1e30", "1e100", "1e250", "1e1000", "1e3000", "1e10000", "1e100000"],
  //Fire
  ["1000", "1e6", "1e12", "1e18", "1e30", "1e100", "1e250", "1e1000", "1e4000", "1e20000"],
  //Platinum
  ["100", "100000", "2.5e6", "1e8", "1e12", "1e20", "1e30", "1e100", "1e250", "1e1000"],
  //Magic
  ["1", "5", "50", "1000", "100000", "1e8", "1e20", "1e100", "1e500", "1e10000", "1e50000", "1e1000000"],
  //Magifolds
  ["10000", "1e6", "1e20", "1e30", "1e50", "1e250", "1e1000", "1e2500", "1e10000"],
  //Uranium
  ["100", "100000", "1e7", "5e8", "1e10", "1e100", "1e1000"],
  //Cyan sigils
  ["1", "5", "20", "100", "500", "1000000", "1e12"],
  //Blue sigils
  ["1", "5", "20", "100", "500", "1000000", "1e12"],
  //Indigo sigils
  ["1", "5", "20", "100", "500", "1000000", "1e12"],
  //Violet sigils
  ["1", "5", "20", "100", "500", "1000000", "1e12"],
  //Pink sigils
  ["1", "5", "20", "100", "500", "1000000", "1e12"],
  //Knowledge
  ["1", "10", "250", "10000", "1e9"],
  //Tomes
  ["1", "10"]
]
const achievementRewards = {
  //gold
  ach0x4: "Unlocks Max All button for platinum upgrades!",
  ach0x8: "Unlocks autobuyer for miners! Also allows you to keep 1 miner when resetting.",
  ach0x10: "Unlock Buy All buttons for magic and dark magic upgrades!",
  //magic
  ach3x0: "Unlocks Buy Max button for fire upgrades!",
  ach3x1: "Unlocks Max Buy All button for fire upgrades!",
  ach3x4: "Keep all platinum upgrades on magic reset!",
  ach3x5: "Gain 1% of potential magic per second!",
  ach3x7: "Gain 100% of potential magic per second!",
  //uranium
  ach5x2: "Unlocks Max All button for uranium upgrades!",
  ach5x3: "Automatically gain uranium without resetting!",
  //cyan sigils
  ach6x0: "Automatically spend time with your dragon when available! Also allows you to enter challenges while on cooldown.",
  ach6x1: "Feeding your dragon reduces scores by ^0.5 instead of resetting them!",
  ach6x2: "Automatically max all fire upgrades!",
  //blue sigils
  ach7x0: "Sigil resets reduce scores by ^0.5 instead of resetting them!",
  ach7x1: "Feeding your dragon no longer reduces challenge scores!",
  ach7x2: "Keep all platinum upgrades on sigil reset!",
  //indigo sigils
  ach8x0: "Gain challenge score without entering challenges! Also, sigil resets no longer reduce challenge scores.",
  ach8x1: "Keep uranium upgrades on all resets!",
  //violet sigils
  ach9x0: "Unlocks the ability to pet your dragon! :O",
  //pink sigils
  ach10x0: "Automatically feed your dragon when available!",
  ach10x1: "Unlocks Max All button for sigil upgrades!",
  ach10x2: "Keep all magic and dark magic upgrades on sigil reset!",
  //knowledge
  ach11x0: "Unlocks some automation for sigil resets!"
}
const achievementResources = [
  {name:"gold", shortName:"&#8202;&#8202;Gold", internalName:"gold"},
  {name:"fire", shortName:"&#8202;&#8202;Fire", internalName:"fire"},
  {name:"platinum", shortName:"Platinum", internalName:"platinum"},
  {name:"magic", shortName:"Magic", internalName:"magic"},
  {name:"magifolds", shortName:"Magifolds", internalName:"magifolds"},
  {name:"uranium", shortName:"Uranium", internalName:"uranium"},
  {name:"cyan sigils", shortName:"C-sigils", internalName:"cyanSigils"},
  {name:"blue sigils", shortName:"&#8202;&#8202;B-sigils", internalName:"blueSigils"},
  {name:"indigo sigils", shortName:"&#8202;&#8202;I-sigils", internalName:"indigoSigils"},
  {name:"violet sigils", shortName:"&#8202;&#8202;V-sigils", internalName:"violetSigils"},
  {name:"pink sigils", shortName:"P-sigils", internalName:"pinkSigils"},
  {name:"knowledge", shortName:"Knwlege", internalName:"knowledge"},
  {name:"total tomes", shortName:"Tomes", internalName:"totalTomes"}
]
const achievementColours = ["#fd0", "#f80", "#bff", "#90d", "#60d", "#8f8", "#06b", "#00b", "#40b", "#60b", "#b0b", "#987", "#965"]
//this array tells the display code when each achievement should be visible. each child array corresponds to a value of game.unlocks
const achievementDisplayUnlocks = [
  [3,2,1,0,0,0,0,0,0,0,0,0,0], //unlock 0 - start
  [3,2,1,0,0,0,0,0,0,0,0,0,0], //unlock 1 - dragon
  [3,2,1,0,0,0,0,0,0,0,0,0,0], //unlock 2 - fire
  [4,3,2,1,0,0,0,0,0,0,0,0,0], //unlock 3 - platinum
  [6,4,3,5,1,0,0,0,0,0,0,0,0], //unlock 4 - magic
  [6,4,3,5,1,0,0,0,0,0,0,0,0], //unlock 5 - magic challenges
  [7,5,4,6,2,1,0,0,0,0,0,0,0], //unlock 6 - more magic upgrades
  [8,6,6,7,3,1,0,0,0,0,0,0,0], //unlock 7 - uranium
  [10,6,7,8,4,3,1,0,0,0,0,0,0], //unlock 8 - more platinum and uranium
  [10,6,7,8,4,3,1,0,0,0,0,0,0], //unlock 9 - dark magic
  [10,6,7,9,4,4,3,1,0,0,0,0,0], //unlock 10 - cyan
  [11,7,8,9,5,4,4,4,1,0,0,0,0], //unlock 11 - blue
  [12,8,8,10,6,5,5,5,5,1,0,0,0], //unlock 12 - indigo
  [12,8,9,11,6,5,5,5,5,5,1,0,0], //unlock 13 - violet
  [12,9,9,11,7,5,5,5,5,5,5,1,0], //unlock 14 - pink
  [12,10,10,12,8,6,6,6,6,6,6,4,1], //unlock 15 - knowledge
  [12,10,10,12,9,7,7,7,7,7,7,5,2], //unlock 16 - tomes
]